torch/ao/quantization/dynamic_tracing/auto_trace.py

```diff
@@ -0,0 +1,554 @@
+        quantization.
+
+        For each function with a `__torch_function__` override, this proxy does
+        the following for functions which need quantization:
+
+        1. calls `_auto_quant_state.validate_cur_op` to validate that
+           the currently seen op is the same as what was recorded during tracing
+        2. calls `_auto_quant_state.op_prepare_before_hook`
+        3. executes the original function
+        4. calls `_auto_quant_state.op_prepare_after_hook`
+        5. calls `_auto_quant_state.mark_cur_op_complete` to increment
+           the current op index in preparation for the next op
+
+        Otherwise, calls the original function.
+        """
+
+        @classmethod
+        def __torch_function__(cls, func, types, args=(), kwargs=None):
+            # to prevent printing things from going into an infinite loop
+            if func == torch.Tensor.__repr__:
+                return super().__torch_function__(func, types, args, kwargs)
+            if enable_logging:
+                logger.debug(f'__torch_function__ {str(func)} len_args {len(args)}')
+
+            nonlocal qtensor_id
+            nonlocal cur_module
+            kwargs = kwargs if kwargs else {}
+            # if we are in a function, the current module is always a parent
+            parent_module = cur_module
+            hook_type = get_torch_function_hook_type(parent_module, func)
+
+            if hook_type is HookType.OP_HOOKS:
+                assert parent_module is not None
+                qstate = parent_module._auto_quant_state
+                fqn = module_id_to_fqn[id(parent_module)] if parent_module else None
+                if not first_call:
+                    qstate.validate_cur_op(func)
+                # run "before" hook
+                args, kwargs = qstate.op_prepare_before_hook(
+                    func, args, kwargs, first_call, qtensor_id, fqn, parent_module)
+                # forward
+                output = super().__torch_function__(func, types, args, kwargs)
+                # run "after" hook
+                output = qstate.op_prepare_after_hook(
+                    func, output, args, first_call, qtensor_id, parent_module)
+                qstate.mark_cur_op_complete(func)
+            else:
+                output = super().__torch_function__(func, types, args, kwargs)
+
+            # TODO: is this right? Don't really understand this
+            if output is NotImplemented:
+                with torch._C.DisableTorchFunction():
+                    output = func(*args, **kwargs).as_subclass(
+                        QuantizationInterceptionProxy)
+                assert output is not NotImplemented
+
+            return output
+
+        def __repr__(self):
+            return f'QuantizationInterceptionProxy({super().__repr__()})'
+
+        # TODO(future PR): add other math overrides
+
+    class QuantizationInterceptionModule(type(model)):  # type: ignore[misc]
```

Comment:
What is the difference in using `type(model)` vs explicit `nn.Module` here?

Reply:
if we cast to `nn.Module`, then this would override any API on the original model which doesn't exist on `nn.Module`